packages/apps/src/microsoft/teams/apps/app.py

```diff
@@ -333,18 +336,74 @@ async def _refresh_graph_token(self, force: bool = False) -> None:
             self.log.debug("Refreshing graph token")
 
         try:
-            # TODO: Implement graph token refresh when graph client is available
-            # token_response = await self.api.bots.token.get_graph(self.credentials)
-            # self._tokens.graph = TokenProtocol(token_response.access_token)
-            self.log.debug("Graph token refresh not yet implemented")
+            # Get graph token using the API client
+            token_response = await self.api.bots.token.get_graph(self.credentials)
+            self._tokens.graph = JsonWebToken(token_response.access_token)
+
+            self.log.debug("Graph token refreshed successfully")
+
+            # Emit token acquired event
+            self._events.emit("token", {"type": "graph", "token": self._tokens.graph})
 
-            # When implemented, emit token event:
         except Exception as error:
             self.log.error(f"Failed to refresh graph token: {error}")
 
             self._events.emit("error", ErrorEvent(error, context={"method": "_refresh_graph_token"}))
             raise
 
+    async def _get_or_refresh_graph_token(self) -> Optional[TokenProtocol]:
+        """Get the current graph token, refreshing if necessary."""
+        try:
+            await self._refresh_graph_token()
+            return self._tokens.graph
+        except Exception as e:
+            self.log.error(f"Failed to get or refresh graph token: {e}")
+            return self._tokens.graph  # Return current token even if refresh failed
+
+    async def get_or_refresh_tenant_token(self, tenant_id: Optional[str] = None) -> Optional[str]:
+        """Get or refresh a tenant-specific graph token, mirroring TypeScript getOrRefreshTenantToken."""
+        if not tenant_id:
+            # Use default tenant from credentials if no specific tenant provided
```

Comment:
This is a bit dangerous. Do we do this in typescript? If we do we shouldn't. This is an opportunity for data leaks between tenants